src/main/java/software/amazon/nio/spi/s3/S3ClientProvider.java

```diff
@@ -0,0 +1,288 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package software.amazon.nio.spi.s3;
+
+import java.io.IOException;
+import java.net.URI;
+import java.time.Duration;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.awscore.AwsClient;
+import software.amazon.awssdk.core.exception.ApiCallAttemptTimeoutException;
+import software.amazon.awssdk.core.exception.ApiCallTimeoutException;
+import software.amazon.awssdk.core.exception.RetryableException;
+import software.amazon.awssdk.core.retry.backoff.EqualJitterBackoffStrategy;
+import software.amazon.awssdk.core.retry.conditions.OrRetryCondition;
+import software.amazon.awssdk.core.retry.conditions.RetryCondition;
+import software.amazon.awssdk.core.retry.conditions.RetryOnClockSkewCondition;
+import software.amazon.awssdk.core.retry.conditions.RetryOnExceptionsCondition;
+import software.amazon.awssdk.core.retry.conditions.RetryOnStatusCodeCondition;
+import software.amazon.awssdk.core.retry.conditions.RetryOnThrottlingCondition;
+import software.amazon.awssdk.http.HttpStatusCode;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.S3ClientBuilder;
+import software.amazon.awssdk.services.s3.S3CrtAsyncClientBuilder;
+import software.amazon.awssdk.services.s3.model.HeadBucketResponse;
+import software.amazon.awssdk.services.s3.model.S3Exception;
+import software.amazon.nio.spi.s3.config.S3NioSpiConfiguration;
+
+/**
+ *
+ */
+public class S3ClientProvider {
+
+    /**
+     * Default S3CrtAsyncClientBuilder
+     */
+    protected S3CrtAsyncClientBuilder asyncClientBuilder = S3AsyncClient.crtBuilder();
+
+
+    /**
+     * Default client using the "https://s3.us-east-1.amazonaws.com" endpoint
+     */
+    private static final S3Client DEFAULT_CLIENT = S3Client.builder()
+            .endpointOverride(URI.create("https://s3.us-east-1.amazonaws.com"))
+            .region(Region.US_EAST_1)
+            .build();
+
+    /**
+     * Default asynchronous client using the "https://s3.us-east-1.amazonaws.com" endpoint
+     */
+    private static final S3AsyncClient DEFAULT_ASYNC_CLIENT = S3AsyncClient.builder()
+            .endpointOverride(URI.create("https://s3.us-east-1.amazonaws.com"))
+            .region(Region.US_EAST_1)
+            .build();
+
+        private final EqualJitterBackoffStrategy backoffStrategy = EqualJitterBackoffStrategy.builder()
+            .baseDelay(Duration.ofMillis(200L))
+            .maxBackoffTime(Duration.ofSeconds(5L))
+            .build();
+
+    final RetryCondition retryCondition;
```

Comment:
No particular reason, it is coming from the original implementation. I guess it comes from being DEFAULT clients static instances. If we want to improve it, I would suggest to keep it for later.

Reply:
It might actually be good to move the back off strategy and retry strategy objects out to a utility class and make them public static final objects to promote reuse?